### PR TITLE
Adding grep and tail like functionality

### PR DESCRIPTION
**Pull request by dpengftw**
##### SUMMARY

Tail and grep like function in a module

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

greppy

##### ADDITIONAL INFORMATION

Sample ad-hoc invocation of module.

```paste below
 ~/community.general | on main  ANSIBLE_LIBRARY=plugins/modules ansible localhost --connection local -m greppy -a "path=README.md search=code exclude=ansible ignore_cas
e=yes timeout=3"
[WARNING]: Found both group and host with same name: localhost
localhost | SUCCESS => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python3"
    },
    "changed": false,
    "exclude_matches": 43,
    "matches": 1,
    "output": [
        "## Code of Conduct\n"
    ]
}
```

### CODE REVIEW

File: plugins/modules/greppy.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+# SPDX-License-Identifier: GPL-3.0-or-later
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: greppy
+
+short_description: Scan logs for certain patterns within a given timeframe.
```
Comment (felixfontein) [inline plugins/modules/greppy.py:13]:
```suggestion
short_description: Scan logs for certain patterns within a given timeframe

version_added: 6.1.0
```

File: plugins/modules/greppy.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+# If this is part of a collection, you need to use semantic versioning,
+# i.e. the version is of the form "2.5.0" and not "2.4".
+
```
Comment (felixfontein) [inline plugins/modules/greppy.py:17]:
```suggestion
```

Please remove such comments. They do not really provide value for actual modules.

File: plugins/modules/greppy.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        required: true
+        type: list
+        elements: str
+    exclude:
+        description: The literal string to exclude from the file. This does not have to match the entire line.
+        required: false
+        type: list
+        elements: str
+    timeout:
+        description: The number of seconds to wait.
```
Comment (felixfontein) [inline plugins/modules/greppy.py:36]: I would mention why you would want to wait (namely when looking at a file that's still being written).

File: plugins/modules/greppy.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+        required: false
+        default: 60
+        type: int
+    ignore_case:
+        description: Case insensitive regex search.
+        required: false
+        type: bool
+    findexit:
```
Comment (felixfontein) [inline plugins/modules/greppy.py:44]: How about calling this option `find_only_first` or something like that? `findexit` sounds strange to me. Or is this modelled after another program?

File: plugins/modules/greppy.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+        description: Exit right after finding the first occurrence
+        required: false
+        default: false
+        type: bool
+# Specify this value according to your collection
+# in format of namespace.collection.doc_fragment_name
+extends_documentation_fragment:
+    - community.general.greppy
```
Comment (felixfontein) [inline plugins/modules/greppy.py:52]:
```suggestion
```
You are not adding a documentation fragment of that name, so don't reference it.

File: plugins/modules/greppy.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    search:
+      - first string
+    timeout: 120
+
+- name: Grep for first occurrence then exit
+  community.general.greppy:
+    path: example.txt
+    search:
+      - first occurrence
+    find_exit: yes
```
Comment (felixfontein) [inline plugins/modules/greppy.py:81]:
```suggestion
    find_exit: true
```
Always use `true` and `false` for YAML booleans. Never use `True`, `yes`, ...

File: plugins/modules/greppy.py [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+'''
+
+RETURN = '''
+# These are examples of possible return values, and in general should use other names for return values.
+output:
+    description: List of lines that matched given search
```
Comment (felixfontein) [inline plugins/modules/greppy.py:88]:
```suggestion
    description: List of lines that matched given search.
```

File: plugins/modules/greppy.py [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+  community.general.greppy:
+    path: example.txt
+    search:
+      - first occurrence
+    find_exit: yes
+
+'''
+
+RETURN = '''
+# These are examples of possible return values, and in general should use other names for return values.
```
Comment (felixfontein) [inline plugins/modules/greppy.py:86]:
```suggestion
```

File: plugins/modules/greppy.py [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+    description: List of lines that matched given search
+    type: list
+    returned: always
+    sample: [
+            "Ex eorum enim scriptis et institutis cum omnis doctrina liberalis, omnis historia.\n",
+            "Esse enim quam vellet iniquus iustus poterat inpune.\n"
+        ]
+
+matches:
+    description: Number of lines that matched
```
Comment (felixfontein) [inline plugins/modules/greppy.py:97]:
```suggestion
    description: Number of lines that matched.
```

File: plugins/modules/greppy.py [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+    type: int
+    returned: always
+    sample: 2
+
+exclude_matches:
+    description: Number of lines that matched the exclusion strings
```
Comment (felixfontein) [inline plugins/modules/greppy.py:103]:
```suggestion
    description: Number of lines that matched the exclusion strings.
```

File: plugins/modules/greppy.py [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+from ansible.module_utils._text import to_bytes, to_native
+
+def run_checks(module):
+    """Do baseline checks before proceeding."""
+    params = module.params
+    path = params['path']
+
+    b_dest = to_bytes(path, errors='surrogate_or_strict')
+    if os.path.isdir(b_dest):
+        module.fail_json(rc=256, msg='Destination %s is a directory !' % path)
```
Comment (felixfontein) [inline plugins/modules/greppy.py:122]: That space should only be there in French (and maybe other languages), but not in English.
```suggestion
        module.fail_json(rc=256, msg='Destination %s is a directory!' % path)
```
Another one further below.

File: plugins/modules/greppy.py [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+        time_counter += 1
+        if time_counter > time_to_wait:
+            module.fail_json(rc=256, msg='Destination %s does not exist !' % path)
+
+def search_line(pattern, string):
+    """Find pattern match in string."""
+    if type(string) is bytes:
+        pc = re.compile(bytes(pattern, "utf-8"))
+    else:
+        pc = re.compile(pattern)
```
Comment (felixfontein) [inline plugins/modules/greppy.py:137]:
```suggestion
    pc = re.compile(to_text(pattern))
```
where `from ansible.module_utils.common.text.converters import to_text`.

File: plugins/modules/greppy.py [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+
+    if re.search(pc, string):
+        return string
+    else:
+        return None
+
+def grep_file(filename, patterns, excludes, ignore_case, timeout, find_exit):
+    """Search a single file."""
+    file = open(filename, 'rb')
```
Comment (felixfontein) [inline plugins/modules/greppy.py:146]: Where are you closing this file?

File: plugins/modules/greppy.py [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+    result = {
+        'output': [],
+        'exclude_matches': 0,
+        'matches': 0
+    }
+
+    timeout_start = time.time()
+
+    while time.time() < timeout_start + timeout:
+        line = file.readline()
```
Comment (felixfontein) [inline plugins/modules/greppy.py:156]: If reading a line blocks, how do you control the timeout?

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! I've added some first comments.

Please note that all new modules must have tests (unit tests and/or integration tests).

File: plugins/modules/greppy.py (hunk 4 above)
Comment (dpengftw) [reply plugins/modules/greppy.py:43]: I like that suggestion, incorporated.

File: plugins/modules/greppy.py (hunk 13 above)
Comment (dpengftw) [reply plugins/modules/greppy.py:141]: Added to bottom of function.

File: plugins/modules/greppy.py (hunk 3 above)
Comment (dpengftw) [reply plugins/modules/greppy.py]: Updated description for timeout.

File: plugins/modules/greppy.py (hunk 14 above)
Comment (felixfontein) [reply plugins/modules/greppy.py:156]: You can always use non-blocking IO and `select`, but that's a lot more low-level. Or use multi-threading. That will definitely make life harder for you though.

**Comment by felixfontein** [comment]
ping @dpengftw 

needs_info

**Pull request closed without merging**